### PR TITLE
Linked servers support

### LINKED ISSUE #1125: MSSQL Linked Servers support
Hi,

I walk through documentation and didn't found any information about support of the Linked Servers in MSSQL.
Only thing that I found and using is extension method DatabaseName that is generating -SQL like this:
```sql
FROM [MyDatabase].[dbo].[User] 
```
that is completely ok while query consist of tables located on one server. But when I need join few tables located on different servers and each server has link to each other via MSSQL Linked Server object I would expect similar method that will allow me to define specific linked server name like this:
```cs
DatabaseName("MyDatabase").ServerName("MyServer")
```
that will generate T-SQL query like this:
```sql
FROM [MyServer].[MyDatabase].[dbo].[User]
```

### PR DESCRIPTION
**Pull request by MaceWindu**
Fixes #1125 

Adds support for linked servers for SQL Server, Informix, SAP HANA and Oracle.

**List of Changes**

This PR adds server name parameter or property to all places where we work with table name components (table, schema, database names).

To preserve compatibility with existing code, server name parameter added as very last optional parameter to all  applicable non-async public API. For async public API, parameter added before cancellation token, so it will require changes to calls, that pass cancellation token parameter.

For internal API, server name parameter added to a logical position to be in line with existing table/schema/database parameters.

- generation of fully-qualified object name with server name for providers, that support linked servers: SQL Server, Informix, SAP HANA, Oracle
- new T4 option `string ServerName`. Setting this property to non-null value will enable generation of mapping attributes with `Server` property, initialized by provided value from `ServerName` option

**Public API changes**
New members:
- `string BulkCopyOptions.ServerName` property
- `string ITable<T>.ServerName` property
- `Table<T>.ChangeServerName(string)` method
- `string IEntityChangeDescriptor.ServerName` property
- `string TableAttribute.Server` property
- `string StoredProcedureAttribute.Server` property
- `EntityMappingBuilder<T>.HasServerName(string)` method
- `string EntityDescriptor.ServerName` property
- `string Sql.TableFunctionAttribute.Server` property
- `Sql.TableQualification.ServerName` enum field
- `string TempTable<T>.ServerName` property
- `ITableMutable<T>.ChangeServerName(string)` method
- `LinqExtensions.ServerName<T>(ITable<T>, string)` method

New parameter in non-async public API (`string serverName = null` as last parameter):
- `DataConnectionExtensions.Merge()`
- `TempTable<T>.ctor()`
- `TempTable<T>.CreateTempTable()`
- `DataExtensions.Insert()`
- `DataExtensions.InsertOrReplace()`
- `DataExtensions.Insert*WithIdentity()`
- `DataExtensions.Update()`
- `DataExtensions.Delete()`
- `DataExtensions.CreateTable()`
- `DataExtensions.DropTable()`

New parameter in async public API (`string serverName = null` as second to last parameter, before token):
- `DataConnectionExtensions.MergeAsync()`
- `DataExtensions.InsertAsync()`
- `DataExtensions.InsertOrReplaceAsync()`
- `DataExtensions.Insert*WithIdentityAsync()`
- `DataExtensions.UpdateAsync()`
- `DataExtensions.DeleteAsync()`
- `DataExtensions.CreateTableAsync()`
- `DataExtensions.DropTableAsync()`

**Internal Changes**
New members:
- `ConvertType.NameToServer` enum field
- `string Sql.TableHelper.ServerName` property
- `string BasicSqlBuilder.GetTableServerName(SqlTable)` method
- `string SqlTable.Server` property

New `string serverName` parameter:
- `QueryRunner.CreateTable<T>.Query()`
- `QueryRunner.CreateTable<T>.QueryAsync()`
- `QueryRunner.Delete<T>.CreateQuery()`
- `QueryRunner.Delete<T>.Query()`
- `QueryRunner.Delete<T>.QueryAsync()`
- `QueryRunner.DropTable<T>.Query()`
- `QueryRunner.DropTable<T>.QueryAsync()`
- `QueryRunner.Insert<T>.CreateQuery()`
- `QueryRunner.Insert<T>.Query()`
- `QueryRunner.Insert<T>.QueryAsync()`
- `QueryRunner.InsertOrReplace<T>.CreateQuery()`
- `QueryRunner.InsertOrReplace<T>.Query()`
- `QueryRunner.InsertOrReplace<T>.QueryAsync()`
- `QueryRunner.InsertWithIdentity<T>.CreateQuery()`
- `QueryRunner.InsertWithIdentity<T>.Query()`
- `QueryRunner.InsertWithIdentity<T>.QueryAsync()`
- `QueryRunner.Update<T>.CreateQuery()`
- `QueryRunner.Update<T>.Query()`
- `QueryRunner.Update<T>.QueryAsync()`
- `ISqlBuilder.ConvertTableName()`
- `ISqlBuilder.BuildTableName()`

### CODE REVIEW

**Comment by MaceWindu** [comment]
Discussed it a bit internally:
- we will change position of serverName parameter for some API's to be placed in logical place for that API (source-level breaking change)
- because of API breaking changes we will move it to 3.0 release
- create branch for 3.0 release so people could build it from source if they need new 3.0 features already

File: Source/LinqToDB/Data/DataConnectionExtensions.cs [hunk 1]
```diff
@@ -1639,6 +1652,7 @@ public static BulkCopyRowsCopied BulkCopy<T>(this ITable<T> table, IEnumerable<T
 
 			return dataConnection.DataProvider.Merge(dataConnection, predicate, true, source.Where(predicate),
 				tableName    ?? table.TableName,
+				serverName   ?? table.ServerName,
```
[the remark this answers is not in the document]
Comment (MaceWindu) [reply Source/LinqToDB/Data/DataConnectionExtensions.cs]: No, it is correct positioning. But this specific API - I hope we can finally remove it in 3.0 as it is old Merge API with a lot of bugs and limitations. Just need to persuade Igor that he don't need it anymore :-)

**Comment by MaceWindu** [comment]
@sdanyliv , @igor-tkachev PR updated to recent version

File: Source/LinqToDB/Data/DataConnectionExtensions.LegacyMerge.cs [hunk 2]
```diff
@@ -382,10 +405,11 @@ private static bool MergeWithUpdate<T>(ITable<T> table)
 			string                   tableName         = null,
 			string                   databaseName      = null,
 			string                   schemaName        = null,
-			CancellationToken        cancellationToken = default)
+			CancellationToken        cancellationToken = default,
+			string                   serverName        = null)
```
Comment (sdanyliv) [inline Source/LinqToDB/Data/DataConnectionExtensions.LegacyMerge.cs]: I would prefer token to be the last parameter.

File: Source/LinqToDB/DataExtensions.cs [hunk 3]
```diff
@@ -544,11 +564,12 @@ public static int Delete<T>([NotNull] this IDataContext dataContext, T obj, stri
 			string            statementHeader = null,
 			string            statementFooter = null,
 			DefaultNullable   defaultNullable = DefaultNullable.None,
-			CancellationToken token           = default)
+			CancellationToken token           = default,
+			string            serverName      = null)
```
Comment (sdanyliv) [inline Source/LinqToDB/DataExtensions.cs]: Also here, token as a last parameter (many places spotted)

File: Source/LinqToDB/DataProvider/DB2/DB2SqlBuilderBase.cs [hunk 4]
```diff
@@ -251,7 +251,7 @@ public override StringBuilder BuildTableName(StringBuilder sb, string database,
 			if (database != null && schema == null)
 				throw new LinqToDBException("DB2 requires schema name if database name provided.");
 
-			return base.BuildTableName(sb, database, schema, table);
+			return base.BuildTableName(sb, null, database, schema, table);
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/DB2/DB2SqlBuilderBase.cs]: Why servername is null here?

File: Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLSqlBuilder.cs [hunk 5]
```diff
@@ -301,7 +302,7 @@ public override StringBuilder BuildTableName(StringBuilder sb, string database,
 			if (database != null && schema == null)
 				database = null;
 
-			return base.BuildTableName(sb, database, schema, table);
+			return base.BuildTableName(sb, null, database, schema, table);
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/PostgreSQL/PostgreSQLSqlBuilder.cs]: Why null?

File: Source/LinqToDB/Data/DataConnectionExtensions.LegacyMerge.cs (hunk 2 above)
Comment (MaceWindu) [reply Source/LinqToDB/Data/DataConnectionExtensions.LegacyMerge.cs]: Me too, but as we discussed before - we don't want to break source compatibility here.

File: Source/LinqToDB/DataProvider/DB2/DB2SqlBuilderBase.cs (hunk 4 above)
Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/DB2/DB2SqlBuilderBase.cs]: all buildtable methods use only supported fragments, DB2 and pgsql don't support server name

**Comment by MaceWindu** [comment]
added overloads with extra parameter for async APIs

**Pull request merged**